### PR TITLE
kvserver: add cross-zone snapshot byte metrics to StoreMetrics

### PR DESCRIPTION
**Pull request by wenyihu6**
**kvserver: refactor getSnapshotBytesMetrics**

This commit refactors `getSnapshotBytesMetrics` function in
`replica_learner_test`. Instead of rigidly defining an array of metrics
information to be extracted within the function, the refactoring now allows the
caller to pass a metrics slice array to retrieve specific metrics of the
caller’s choice.

This commit does not change any existing functionality, and the main purpose is
to make future commits cleaner.

Part of: https://github.com/cockroachdb/cockroach/issues/103983 
Release note: none

---
**kvserver: add cross-zone snapshot byte metrics to StoreMetrics**

Previously, we [added](#104111) cross-region snapshot byte metrics to track the
aggregate of snapshot bytes sent from and received at a store across different
regions. We should add metrics to track cross-zone snapshots as well.

To improve this issue, this commit adds two new store metrics -
```
range.snapshots.cross-zone.sent-bytes 
range.snapshots.cross-zone.rcvd-bytes 
```
These metrics track the aggregate of snapshot bytes sent from and received at a
store across different zones within the same region if the zone and region tier
keys are properly configured across nodes.
	
To ensure accurate metrics and consistent error logging, it is important to
follow the assumption when configuring locality tiers across nodes:
1. For cross-region metrics, region tier keys must be present.
2. For cross-zone metrics, zone tier keys must be present. It is also essential
to maintain consistency in the zone tier key across nodes. (e.g. using different
keys, such as “az” and “zone”, can lead to unexpected behavior).
3. Within a node locality, region and zone tier keys should be unique.
4. Ensure consistent configuration of region and zone tiers across nodes.
5. Cross-region but same zone activities should be impossible. 

If all nodes configure both region and zone tiers:
Cross-region and cross-zone metrics can be used to track following information:
a. Aggregate of cross-region, cross-zone activities:
`range.snapshots.cross-region.(sent|received)-bytes`
b. Aggregate of same-region, cross-zone activities:
`range.snapshots.cross-zone.(sent|received)-bytes`
c. Aggregate of same-region, same-zone activities:
`range.snapshots.(sent|received)-bytes` - a - b
d. Cross-region, same zone activities will be considered as misconfigured, and
an error will be logged.

If all nodes have zone tiers configured, but not regions: cross-zone metrics
will still be accurate. Problems arise if some nodes have region tier keys
configured while others do not.

If all nodes have region tiers configured: cross-region metrics will be accurate
regardless of the nodes’ zone tier setup.

Part of: https://github.com/cockroachdb/cockroach/issues/103983

Release note (ops change): Two new store metrics -
`range.snapshots.cross-zone.sent-bytes` and
`range.snapshots.cross-zone.rcvd-bytes` - are now added to track the aggregate
of snapshot bytes sent from and received at a store across different zones.

For accurate metrics, follow these assumptions:
- Configure region and zone tier keys consistently across nodes.
- Within a node locality, ensure unique region and zone tier keys.
- Maintain consistent configuration of region and zone tiers across nodes.

### CODE REVIEW

**Review by kvoli — COMMENTED** [review]
> Reviewed 1 of 1 files at r1, 7 of 7 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/104417)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @wenyihu6)

___
*[`-- commits` line 29 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/104417#-NXMqOyU0AXP6aGr1l1-):*
Could you mention the locality-tier keys which are recognized as a "zone" here.
___
*[`-- commits` line 35 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/104417#-NXMqD3d9kOUVhCtoxRj):*
Could we shorten the release note and move some of this up above. 

I think it suffices to mention the supported zone locality-tier keys and then the requirement that there be at most one and consistent across nodes. The example is good!
___
*[`pkg/kv/kvserver/store_snapshot.go` line 977 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/104417#-NXMrS4r-3GKbLh47B3X:-NXMrS4r-3GKbLh47B3Y:bgilhe0) ([raw file](https://github.com/cockroachdb/cockroach/blob/21d4babc2ed64ff229c90a5388fcd20b23391779/pkg/kv/kvserver/store_snapshot.go#L977)):*
> ```Go
> // shouldIncrementCrossLocalitySnapshotMetrics returns (bool, bool) - indicating
> // if the two given replicas are cross-region and cross-zone respectively.
> func (s *Store) shouldIncrementCrossLocalitySnapshotMetrics(
> ```

nit: It may be worth mentioning why this function deals with both x-region and x-zone are handled in the same method, perf - so that in the future someone doesn't increase the overhead because they were unaware.
___
*[`pkg/roachpb/metadata.go` line 622 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/104417#-NXMup015KbOeo6tQMxI:-NXMup015KbOeo6tQMxJ:bv8c40i) ([raw file](https://github.com/cockroachdb/cockroach/blob/21d4babc2ed64ff229c90a5388fcd20b23391779/pkg/roachpb/metadata.go#L622)):*
> ```Go
> // getFirstRegionFirstZone iterates through the locality tiers and returns
> // multiple values containing:
> // 1.The value of the first encountered "region" tier.
> ```

nit: add some spacing between "1.The..." 
___
*[`pkg/roachpb/metadata.go` line 628 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/104417#-NXMv2kd111slVm4H6Mu:-NXMv2kd111slVm4H6Mv:b-dt6l1q) ([raw file](https://github.com/cockroachdb/cockroach/blob/21d4babc2ed64ff229c90a5388fcd20b23391779/pkg/roachpb/metadata.go#L628)):*
> ```Go
> func (l Locality) getFirstRegionFirstZone() (string, bool, string, string, bool) {
> 	hasRegion, hasZone := false, false
> 	firstRegionValue, firstZoneKey, firstZoneValue := "", "", ""
> ```

nit: probably more idiomatic to do something like

```go
var ( 
    hasRegion, hasZone bool
    firstRegionValue, firstZoneKey, firstZoneValue string
)
```

Or get that for free if you name the function return values like

```go
func (l Locality) getFirstRegionFirstZone()
(firstRegion string, hasRegion bool, ...) {
```
___
*[`pkg/roachpb/metadata.go` line 654 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/104417#-NXMzA_k-vq6gQ00EJWQ:-NXMzA_k-vq6gQ00EJWR:b970xe4) ([raw file](https://github.com/cockroachdb/cockroach/blob/21d4babc2ed64ff229c90a5388fcd20b23391779/pkg/roachpb/metadata.go#L654)):*
> ```Go
> // 2. Error indicating if either locality does not have a "region" tier key.
> // 3. A boolean value indicating if this and the provided locality are
> // cross-region.
> ```

cross-zone*
___
*[`pkg/roachpb/metadata.go` line 667 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/104417#-NXN-MdVAg6Oq-piAsas:-NXN-MdVAg6Oq-piAsat:bbcq7hb) ([raw file](https://github.com/cockroachdb/cockroach/blob/21d4babc2ed64ff229c90a5388fcd20b23391779/pkg/roachpb/metadata.go#L667)):*
> ```Go
> // be checked thoroughly for duplicate zone tier keys and key mismatches.
> // However, due to frequent invocation of this function, we prefer to terminate
> // the check after examining the first encountered zone tier key-value pairs.
> ```

Just checking my understanding. Ignore possibility of any invalid locality keys, so they're all either `region` or `zone`.

Is the intention for x-region to imply x-zone? i.e. should it ever be possible to have two localities which are x-region but not x-zone?

When you want to estimate out how many bytes (roughly) you are paying for at x-region rates (excluding cheaper x-zone), you would do `sum(x-region) - sum(x-zone)`?

If so, do we need to additionally assume clusters won't have a mix of `locality=region=...,zone=...`, `locality=region=...` and `locality=zone=...`?


For example

```
L1: region=us-east, zone=us-east-1
L2:                 zone=us-east-2
L3: region=us-west

L1 x L2 = x-zone
L1 x L3 = x-region
L2 x L3 = none
```

Presume 3 snaps/requests each with value 1 for each of the `Li x Lj` enumerated above. The`sum(x-region)-sum(x-zone)` metric query would return 0 despite there really being 1 x-region and 1 x-zone.

I know we previously chatted about the x-zone condition requiring a `region` flag also be set, what are your thoughts on this?
___
*[`pkg/roachpb/metadata.go` line 668 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/104417#-NXN7osLA6QzPkOs3g8m:-NXN7osLA6QzPkOs3g8n:bqxwxau) ([raw file](https://github.com/cockroachdb/cockroach/blob/21d4babc2ed64ff229c90a5388fcd20b23391779/pkg/roachpb/metadata.go#L668)):*
> ```Go
> // However, due to frequent invocation of this function, we prefer to terminate
> // the check after examining the first encountered zone tier key-value pairs.
> func (l Locality) IsCrossRegionCrossZone(other Locality) (bool, error, bool, error) {
> ```

nit: Consider naming the return values.
___
*[`pkg/roachpb/metadata_test.go` line 224 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/104417#-NXN7bFO9RsyPXhd-O0e:-NXN7bFO9RsyPXhd-O0f:b-9fqeo9) ([raw file](https://github.com/cockroachdb/cockroach/blob/21d4babc2ed64ff229c90a5388fcd20b23391779/pkg/roachpb/metadata_test.go#L224)):*
> ```Go
> }
> 
> func TestLocalityIsCrossRegionCrossZone(t *testing.T) {
> ```

Great test!

**Comment by wenyihu6** [comment]
*[`-- commits` line 29 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/104417#-NXMqOyU0AXP6aGr1l1-):*
<details><summary><i>Previously, kvoli (Austen) wrote…</i></summary><blockquote>

Could you mention the locality-tier keys which are recognized as a "zone" here.
</blockquote></details>

Done.

**Comment by wenyihu6** [comment]
*[`-- commits` line 35 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/104417#-NXMqD3d9kOUVhCtoxRj):*
<details><summary><i>Previously, kvoli (Austen) wrote…</i></summary><blockquote>

Could we shorten the release note and move some of this up above. 

I think it suffices to mention the supported zone locality-tier keys and then the requirement that there be at most one and consistent across nodes. The example is good!
</blockquote></details>

I wasn't sure the level of details we want to include for the commit and release notes. Let me know if this is too wordy.

**Comment by wenyihu6** [comment]
*[`pkg/kv/kvserver/store_snapshot.go` line 977 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/104417#-NXMrS4r-3GKbLh47B3X:-NXR8CBR9e1a9iXW9kPq:bddgjwm) ([raw file](https://github.com/cockroachdb/cockroach/blob/21d4babc2ed64ff229c90a5388fcd20b23391779/pkg/kv/kvserver/store_snapshot.go#L977)):*
<details><summary><i>Previously, kvoli (Austen) wrote…</i></summary><blockquote>

nit: It may be worth mentioning why this function deals with both x-region and x-zone are handled in the same method, perf - so that in the future someone doesn't increase the overhead because they were unaware.
</blockquote></details>

I added a [comment](https://github.com/cockroachdb/cockroach/blob/d49f7d787cf043cc3c413fa97e4d0dea78f16e33/pkg/roachpb/metadata.go#L673C3-L675) for the locality helper instead. I think if someone is looking to add more locality tiers comparison here, they would come across that function eventually. I'm adding `shouldIncrementMetrics` for other PRs as well, so I'm trying to avoid adding this comment redundantly.

**Comment by wenyihu6** [comment]
*[`pkg/roachpb/metadata_test.go` line 224 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/104417#-NXN7bFO9RsyPXhd-O0e:-NXR8mqTB855TseBAMow:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/21d4babc2ed64ff229c90a5388fcd20b23391779/pkg/roachpb/metadata_test.go#L224)):*
<details><summary><i>Previously, kvoli (Austen) wrote…</i></summary><blockquote>

Great test!
</blockquote></details>

Done.

**Comment by wenyihu6** [comment]
*[`pkg/roachpb/metadata.go` line 667 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/104417#-NXN-MdVAg6Oq-piAsas:-NXR8qZc9bnUnpLLmOjS:bb4x8ub) ([raw file](https://github.com/cockroachdb/cockroach/blob/21d4babc2ed64ff229c90a5388fcd20b23391779/pkg/roachpb/metadata.go#L667)):*
<details><summary><i>Previously, kvoli (Austen) wrote…</i></summary><blockquote>

Just checking my understanding. Ignore possibility of any invalid locality keys, so they're all either `region` or `zone`.

Is the intention for x-region to imply x-zone? i.e. should it ever be possible to have two localities which are x-region but not x-zone?

When you want to estimate out how many bytes (roughly) you are paying for at x-region rates (excluding cheaper x-zone), you would do `sum(x-region) - sum(x-zone)`?

If so, do we need to additionally assume clusters won't have a mix of `locality=region=...,zone=...`, `locality=region=...` and `locality=zone=...`?


For example

```
L1: region=us-east, zone=us-east-1
L2:                 zone=us-east-2
L3: region=us-west

L1 x L2 = x-zone
L1 x L3 = x-region
L2 x L3 = none
```

Presume 3 snaps/requests each with value 1 for each of the `Li x Lj` enumerated above. The`sum(x-region)-sum(x-zone)` metric query would return 0 despite there really being 1 x-region and 1 x-zone.

I know we previously chatted about the x-zone condition requiring a `region` flag also be set, what are your thoughts on this?




</blockquote></details>

Discussed this offline: 
1. X-region activities should imply x-zone. Failing to adhere to this assumption leads to error logging. 
2. Consistent zone tier configuration across nodes is also required. See more details in the PR notes. 

In addition, the initial commit always increments cross-zone metrics. 

If we want to calculate for cross zone within same region activities, we would have to use 
`sum(cross-zone within the same region) = sum(cross-zone) - sum(cross-region)`. 

<details>

```
Given the assumption that x-region implies x-zone, sum(cross-zone between different regions) == sum(cross-region). 

sum(cross-zone within the same region) = sum(cross-zone) - sum(cross-zone between different regions)
=> sum(cross-zone within the same region) = sum(cross-zone) - sum(cross-region) by the assumption
```

</details>

The new commit now only increments if it is cross-zone, within-region activities for properly configured region tiers. This makes the metrics more meaningful.

**Review by kvoli — APPROVED** [review]
:lgtm: with some small suggestions.

> Reviewed 4 of 6 files at r3, 2 of 2 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/104417#-:-NXWU5-14JnRw-Dq7W5F:b-c04vtz)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @wenyihu6)

___
*[`-- commits` line 19 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/104417#-NXWRLX50UfRIzcMYp64):*
nit: github will pickup the link fine with just #104111. No need for markdown.
___
*[`-- commits` line 33 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/104417#-NXWRjGsA0ekGuRjAVch):*
nit: Are these commit messages formatted to have correct line length?
___
*[`pkg/kv/kvserver/store_snapshot.go` line 1004 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/104417#-NXWT-kX0w-3XVaPfo_Q:-NXWT-kX0w-3XVaPfo_R:b-87u098) ([raw file](https://github.com/cockroachdb/cockroach/blob/6c70039e055cffc9bc1051d5546b74c23c01d196/pkg/kv/kvserver/store_snapshot.go#L1004)):*
> ```Go
> 	isCrossRegion, isCrossZone := s.shouldIncrementCrossLocalitySnapshotMetrics(ctx, firstReplica, secReplica)
> 	switch sentOrReceived {
> 	case "sent":
> ```

Should this instead be a bool? It would be better to be as strict as possible.

**Comment by wenyihu6** [comment]
*[`pkg/kv/kvserver/store_snapshot.go` line 1004 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/104417#-NXWT-kX0w-3XVaPfo_Q:-NXWYJPBC3fmmZk07w71:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6c70039e055cffc9bc1051d5546b74c23c01d196/pkg/kv/kvserver/store_snapshot.go#L1004)):*
<details><summary><i>Previously, kvoli (Austen) wrote…</i></summary><blockquote>

Should this instead be a bool? It would be better to be as strict as possible.
</blockquote></details>

Done.

**Comment by wenyihu6** [comment]
TFTR!

bors r=kvoli
